File: WebClient/src/ICommunicator.ts
import { IResponse } from "./IResponse";
import { IMessage } from "./IMessage";

export interface ICommunicator {
    //publish message under certain topic
    publish: (topic: string, message: string) => void;
    //subscribe to a topic, store the callback function for that topic, and return a promise of IResponse
    subscribeAsync: (topic: string, topicCallback: (message: IMessage) => any) => Promise<IResponse>;
    //unsubscribe from a topic, remove the cached callback, and return a promise of IResponse
    unsubscribeAsync: (topic: string) => Promise<IResponse>;
}

File: Shared/ICommunicator.cs
using System;
using System.Threading.Tasks;

namespace SignalRCommunicator
{
    public interface ICommunicator
    {
        /// <summary>
        /// Subscribe to a topic to receive the messages from this topic
        /// </summary>
        /// <param name="topic">The name of the topic to be subscribed to</param>
        /// <param name="callBack">Handles the response</param>
        /// <returns></returns>
        Task<IResponse> SubscribeAsync(string topic, Func<string, Task> callBack);

        /// <summary>
        /// Unsuscribe for a topic to stop receiving the messages from this topic
        /// </summary>
        /// <param name="topic">The name of the topic to unsubscribe from</param>
        /// <returns></returns>
        Task<IResponse> UnsubscribeAsync(string topic);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="topic"></param>
        /// <param name="message"></param>
        /// <returns></returns>

        Task PublishAsync(string topic, string message);
    }
}